File: TypeScript/LinkedList.ts
class ListNode<T> {
  value: T
  next: ListNode<T> | null

  constructor(value: T, next: ListNode<T> | null = null) {
    this.value = value
    this.next = next
  }
}

export class LinkedList<T> {
  _head: ListNode<T> | null

  constructor(head: ListNode<T> | null = null) {
    this._head = head
  }

  push(value: T): LinkedList<T> {
    let newNode = new ListNode(value, this._head)
    this._head = newNode

    return this
  }

  map(fn: (value: T) => any): LinkedList<any> {
    if (!this.head()) return new LinkedList<any>()

    let newList = new LinkedList<any>()

    let currentNode = this._head
    while (currentNode) {
      newList.push(fn(currentNode.value))
      currentNode = currentNode.next
    }

    return newList.reverse()
  }

  filter(fn: (a: T) => boolean): LinkedList<T> {
    if (!this.head()) return new LinkedList<T>()

    let newList = new LinkedList<T>()
    let currentNode = this._head
    while (currentNode) {
      if (fn(currentNode.value)) newList.push(currentNode.value)
      currentNode = currentNode.next
    }

    return newList.reverse()
  }

  filter2(fn: (a: T) => boolean): LinkedList<T> {
    if (!this._head) return this

    let currentNode: ListNode<T> | null = this._head;
    let previousNode = this._head
    while (currentNode) {
      if (!fn(currentNode.value)) {
        if (currentNode == this._head)
          this._head = currentNode.next
        else
          previousNode.next = currentNode.next
      }

      previousNode = currentNode
      currentNode = currentNode.next
    }

    return this
  }

  reduce(fn: (a: T, b: T) => T, acc: T): T {
    if (!this.head()) return acc

    let currentNode = this._head
    while (currentNode) {
      acc = fn(currentNode.value, acc)
      currentNode = currentNode.next
    }

    return acc
  }

  reverse(): LinkedList<T> {
    let reversed = new LinkedList<T>()
    let currentNode = this._head

    while (currentNode) {
      reversed.push(currentNode.value)
      currentNode = currentNode.next
    }

    return reversed
  }

  head(): T | null {
    return this._head?.value || null
  }

  tail(): LinkedList<T> {
    if (!this._head) return new LinkedList<T>()

    return new LinkedList<T>(this._head.next)
  }

  index(index: number): T | null {
    if (index < 0 || index >= this.length()) return null

    let currentIndex: number = 0
    let currentNode = this._head

    while (currentIndex < index && currentNode && currentNode.next) {
      currentNode = currentNode.next
      currentIndex++
    }

    return currentNode?.value || null
  }

  last(): T | null {
    let currentNode = this._head
    while (currentNode && currentNode.next) {
      currentNode = currentNode.next
    }

    return currentNode?.value || null
  }

  isEmpty(): boolean {
    return this.length() === 0
  }

  length(): number {
    if (!this.head()) return 0

    let length = 0
    let currentNode = this._head
    while (currentNode) {
      length++
      currentNode = currentNode.next
    }

    return length
  }

  mergeSort(): LinkedList<T> {
    if (this.length() === 1) return this

    const half = Math.floor(this.length() / 2)
    const left = this.take(half)
    const right = this.drop(half)

    return this.merge(left.mergeSort(), right.mergeSort())
  }

  merge(left: LinkedList<T>, right: LinkedList<T>): LinkedList<T> {
    let leftNode = left._head
    let rightNode = right._head
    let newList = new LinkedList<T>()

    while (leftNode && rightNode) {
      if (leftNode.value < rightNode.value) {
        newList.push(leftNode.value)
        leftNode = leftNode.next
      } else {
        newList.push(rightNode.value)
        rightNode = rightNode.next
      }
    }

    while (leftNode) {
      newList.push(leftNode.value)

      leftNode = leftNode.next
    }

    while (rightNode) {
      newList.push(rightNode.value)

      rightNode = rightNode.next
    }

    return newList.reverse()
  }

  take(n: number): LinkedList<T> {
    let newList = new LinkedList<T>()
    if (n === 0) return newList

    let node = this._head
    if (!node) return newList

    for (let i = 0; i < n; i++) {
      newList.push(node.value)
      node = node.next

      if (!node) break
    }

    return newList.reverse()
  }

  drop(n: number): LinkedList<T> {
    let newList = new LinkedList<T>()
    if (n === this.length()) return newList

    let node = this._head
    if (!node) return newList

    for (let i = 0; i < n; i++) {
      node = node.next

      if (!node) break
    }

    while (node) {
      newList.push(node.value)
      node = node.next
    }

    return newList.reverse()
  }
}

File: Haskell/LinkedList.hs
module LinkedList where

import Prelude hiding (filter, head, last, length, map, reverse, tail, drop, take)

data LinkedList a
  = Value a (LinkedList a)
  | Empty

new :: a -> LinkedList a
new value =
  Value value Empty

isEmpty :: LinkedList a -> Bool
isEmpty Empty = True
isEmpty _ = False

head :: LinkedList a -> Maybe a
head Empty = Nothing
head (Value head _) = Just head

tail :: LinkedList a -> LinkedList a
tail (Value _ tail) = tail
tail Empty = Empty

push :: a -> LinkedList a -> LinkedList a
push value list = Value value list

index :: Integer -> LinkedList a -> Maybe a
index 0 list = head list
index idx list
  | idx < 0   = Nothing
  | otherwise = index (idx - 1) (tail list)

length :: LinkedList a -> Integer
length list = lengthAcc list 0
  where
    lengthAcc Empty acc = acc
    lengthAcc (Value listHead listTail) acc = lengthAcc listTail (1 + acc)

reverse :: LinkedList a -> LinkedList a
reverse list = reverseAcc list Empty
  where
    reverseAcc Empty acc = acc
    reverseAcc (Value listHead listTail) acc = reverseAcc listTail (push listHead acc)

last :: LinkedList a -> Maybe a
last Empty = Nothing
last (Value listHead Empty) = Just listHead
last list = last (tail list)

map :: (a -> b) -> LinkedList a -> LinkedList b
map _ Empty = Empty
map fn (Value listHead listTail) = (Value (fn listHead)) (map fn listTail)

filter :: (a -> Bool) -> LinkedList a -> LinkedList a
filter _ Empty = Empty
filter fn (Value listHead listTail)
  | fn listHead = Value listHead (filter fn listTail)
  | otherwise   = filter fn listTail

reduce :: (a -> a -> a) -> a -> LinkedList a -> a
reduce fn acc Empty                     = acc
reduce fn acc (Value listHead listTail) = seq acc' (reduce fn acc' listTail)
                                           where acc' = acc `fn` listHead

mergeSort :: (Ord a) => LinkedList a -> LinkedList a
mergeSort Empty = Empty
mergeSort (Value a Empty) = (Value a Empty)
mergeSort list = merge (mergeSort left) (mergeSort right)
  where
    left = take ((length list) `div` 2) list
    right = drop ((length list) `div` 2) list

merge :: (Ord a) => LinkedList a -> LinkedList a -> LinkedList a
merge left Empty = left
merge Empty right = right
merge (Value head1 tail1) (Value head2 tail2)
  | head1 < head2 = push head1 (merge tail1 right)
  | otherwise     = push head2 (merge left tail2)
    where
      left = (Value head1 tail1)
      right = (Value head2 tail2)

take :: Integer -> LinkedList a -> LinkedList a
take _ Empty = Empty
take 0 list = list
take 1 (Value listHead _) = new listHead
take n (Value listHead listTail) = Value listHead (take (n - 1) listTail)

drop :: Integer -> LinkedList a -> LinkedList a
drop _ Empty = Empty
drop 0 list = list
drop n (Value _ listTail) = drop (n - 1) listTail
